### DISCUSSION TITLE
Why I can search in different languages? How this works? How to understand search, tags and so on?

### DISCUSSION THREAD

**Discussion by aisbergde** [immich-app/immich#6035] [Q&A]
I am trying to understand why I can search in different languages and still get a result. How does the search know which language I'm searching in if I don't specify it? And why do I still get results that match?

Here https://immich.app/docs/features/search it just says:

> Smart search is powered by the pgvecto.rs extension, utilizing machine learning models like CLIP to provide relevant search results. This allows for freeform searches without requiring specific keywords in the image or video metadata.

But from reading here, I still don't understand the whole idea: https://github.com/tensorchord/pgvecto.rs

It looks like some vectors are created, and then photos are compared to vectors. Are vectors assigned with keywords in different languages? Is there an internal translation?

Other questions:

What exactly happens in the "Tag Objects" and "Smart Search" jobs?

And if tags are created, why is there no list of these tags in the GUI? Or is that still planned? Or are there so many tags that it would not be good to display them all in the GUI?

Here is a little bit about the architecture: https://immich.app/docs/developer/architecture
But I still don't understand the basic idea. Can I read into it somewhere? Where to start?

**Comment by alextran1502** [comment]
Object Tagging will be removed in the future because we are not using it for anything. 

For multilingual search, you will need to use a different CLIP model, see available models here https://huggingface.co/immich-app

For search mechanism, we are using CLIP. You can read more here https://openai.com/research/clip

TLDR: powerful blackbox that understands the context of the images

**Reply by aisbergde** [reply]
The default model is "ViT-B-32__openai" (When you do a fresh install and don't change something).

But CLIP and multilingual CLIP models are other models.
There are so many models: hard to start.
I get lost. Is the default a good default? Why it was selected as default?

In General, I understand the idea. I am a cyberneticist by training, and I mainly work with data, databases, data analysis, data mining, prediction, etc. However, I have not yet dealt with CLIP. It seems to be a very interesting topic.

**Reply by mertalev** [reply]
The default is faster and uses less RAM than most of the other models while still performing well. 

For monolingual models, you can check out [this](https://github.com/mlfoundations/open_clip/blob/main/docs/openclip_retrieval_results.csv) spreadsheet that goes into their exact performance. As an example, one of the top performing models (`ViT-H-14-378-quickgelu`) uses about 1054 gigaflops of compute per inference compared to the default model's 14.78.

Multilingual models are generally slower, use more RAM and perform slightly worse for English, but in exchange they can handle over 100 languages. The slowest model we have is `nllb-clip-large-siglip`, using a whopping 1800 gigaflops.

**Reply by aisbergde** [reply]
what about storing not only the result of smart search tagging, but also the used model? This would allow to use different models in parallel and even to switch and compare.

Btw, this would be also possible to use for face tagging. Currently a model switch leads in a loss of all naming and merging and hiding. They could be kept, and this would be a way to coexist with external face tagging.

It's like a tenant in business software.

**Comment by mertalev** [comment]
The model used for smart search is trained on a *huge* amount of data, some of which can be in different languages. It generates a vector for each search query with the goal of encoding its context and meaning. It doesn’t need an explicit sense of “this text is X language”, it just needs to have seen that language during training to understand it.
